### PR TITLE
Implement a global value numbering MIR optimization

### PR DESCRIPTION
**Pull request by cjgillot**
The aim of this pass is to avoid repeated computations by reusing past assignments. It is based on an analysis of SSA locals, in order to perform a restricted form of common subexpression elimination.

By opportunity, this pass allows for some simplifications by combining assignments. For instance, this pass could be able to see through projections of aggregates to directly reuse the aggregate field (not in this PR).

We handle references by assigning a different "provenance" index to each `Ref`/`AddressOf` rvalue. This ensure that we do not spuriously merge borrows that should not be merged. Meanwhile, we consider all the derefs of an immutable reference to a freeze type to give the same value:
```rust
_a = *_b // _b is &Freeze
_c = *_b // replaced by _c = _a
```

### CODE REVIEW

**Comment by cjgillot** [comment]
r? wg-mir-opt

**Review by davidtwco — APPROVED** [review]
This looks good to me but better to get another pair of eyes on it.

**Comment by davidtwco** [comment]
r? wg-mir-opt

**Comment by RalfJung** [comment]
As usual my main question is, what are the high-level assumptions this is making? Those should be documented carefully, ideally at the top of the file so that we can find them again later.

The summary sounds like this is making aliasing model assumptions, specifically reading immutability of data behind `&impl Freeze`. What else is there, e.g. regarding reorderability of operations and things like that?

File: compiler/rustc_mir_transform/src/gvn.rs
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/gvn.rs:437]: I'm not sure that we need to verify that the local is copiable in any sense.

**Comment by cjgillot** [comment]
> The summary sounds like this is making aliasing model assumptions, specifically reading immutability of data behind `&impl Freeze`. What else is there, e.g. regarding reorderability of operations and things like that?

The only thing I added is the immutability behind `&impl Freeze`. I can easily remove it from this PR, to be discussed in a dedicated setting.

I don't assume anything wrt. reorderability: I don't reorder. We only reuse a result if its original computation dominates the reuse. This already allows for quite a few simplifications.

**Comment by RalfJung** [comment]
I am pretty sure this will reorder. If the program does something like
```
let a1 = *a;
let b1 = *b;
let a2 = *a;
```
and you GVN the two `*a` to obtain
```
let a1 = *a;
let b1 = *b;
let a2 = a1;
```
then this effectively means that the second `*a` has been reordered above the `*b`, and then coalesced with the now-neighboring first `*a`. Or did I misunderstand what GVN would do here?

**Comment by cjgillot** [comment]
I had not seen that as reordering. 

My reasoning is simpler. a is &Freeze, so *a is immutable, therefore an execution path from a1 to a2 does not write to *a, which ensures that a1 and a2 hold the same value, and we can unify them. 

Am I missing something ?

**Comment by RalfJung** [comment]
The full argument is more tricky since doing a different sequence of memory operations can also mean you end up with a different state for the aliasing model. It's not enough to just look at the values: in both Stacked Borrows and Tree Borrows, a read has side-effects on the borrow stack/tree.

Of course the aliasing model should allow this optimization, but you need more than just "the value does not change". You also need "not doing the 2nd access will leave the aliasing model in a state that allows the same or more accesses as the original program".

**Comment by cjgillot** [comment]
> You also need "not doing the 2nd access will leave the aliasing model in a state that allows the same or more accesses as the original program".

We already consider that removing reads is fine. For instance in dead-store elimination. The only operation that is never removed is `expose_addr`.

That's why I formulate the argument about immutability. IMO, the reasoning is:
```
a1 = *a;
b1 = *b;
a2 = *a;
```

Becomes
```
a1 = *a;
b1 = *b;
a2 = *a;
assume(a1 == a2); // bytewise comparison of 2 locals, so no aliasing question
a2 = a1;
```

Which is simplified to:
```
a1 = *a;
b1 = *b;
a2 = a1;
```

> I am not sure how far this GVN pass goes

I can definitely make it so it is sufficient. This handling of `&impl Freeze` is the only non-trivial thing this pass does with references and pointers, by making each Ref or AddressOf statement give a new opaque value.

**Comment by RalfJung** [comment]
> That's why I formulate the argument about immutability. IMO, the reasoning is:

Oh, I like that -- that's a nice way to frame this! Makes sense.

**Comment by cjgillot** [comment]
@RalfJung I don't understand what miri is testing in `tests/pass/function_pointers.rs`. This pass makes the assertion `return_fn_ptr(f) != f` fail. We do not guarantee that different reifications of the same function are equal, but do we guarantee that they are unequal?

**Comment by RalfJung** [comment]
No, we don't guarantee either way. That's just testing current Miri behavior, which is kind of a hack.

Ideally this would be non-deterministic. But even then it would be legal for optimizations to make them always come out true/false. So for now it's probably easier if you disable that assertion. Cc @rust-lang/miri

File: src/tools/miri/tests/pass/function_pointers.rs
Comment (RalfJung) [inline src/tools/miri/tests/pass/function_pointers.rs:78]: Let's leave a comment for next time... I would have made this a suggestions but github doesn't let me
```
// Check that `g` always has the same address. This is not guaranteed
// but Miri currently uses a fixed address for monomorphic functions.
(the two lines checking `g`)
// We don't check anything for `f`. Miri gives it many different addresses
// but mir-opts can turn them into the same address.
let _val = return_fn_ptr(f) != f;
```

**Comment by cjgillot** [comment]
This is blocked on https://github.com/rust-lang/rust/issues/112651: we cannot replace a value by a local if there is an implicit subtyping cast in between.

**Comment by cjgillot** [comment]
r? compiler

**Comment by petrochenkov** [comment]
r? compiler

**Comment by wesleywiser** [comment]
r? wesleywiser

**Review by jackh726 — COMMENTED** [review]
Agh sorry.

I started reviewing this, but didn't even really get into the meat of it yet. It was also going to take me a while to grok this code, which I'm not sure familiar with.

Anyways, here's my one small review :)

**Review by oli-obk — APPROVED** [review]
oops had done a review yesterday but seem to have gotten distracted in the middle of it

File: compiler/rustc_mir_transform/src/gvn.rs
Comment (Rageking8) [inline compiler/rustc_mir_transform/src/gvn.rs:4]:
```suggestion
//! such redundancies and re-use the already-computed result when possible.
```

File: compiler/rustc_mir_transform/src/gvn.rs
Comment (Rageking8) [inline compiler/rustc_mir_transform/src/gvn.rs:8]:
```suggestion
//! `Value` is interned as a `VnIndex`, which allows us to cheaply compute identical values.
```

File: compiler/rustc_mir_transform/src/gvn.rs
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/gvn.rs:421]: This is unsound due to https://github.com/rust-lang/rust/issues/115877#issuecomment-1729880362
I need to filter out `ConstValue::Slice` and `Valtree`s as they may contain pointers.

Comment (RalfJung) [reply compiler/rustc_mir_transform/src/gvn.rs:421]: What exactly could go wrong? Is it something like this?
```
const S: &str = "hello";
let x = S;
let y = x; // gets replaced by `S`, no longer guaranteed to be the same pointer as `x`
```

Comment (cjgillot) [reply compiler/rustc_mir_transform/src/gvn.rs:450]: Yes. That's my concern.

Comment (RalfJung) [reply compiler/rustc_mir_transform/src/gvn.rs:450]: I see. Makes sense. We currently do indeed not guarantee that each use of a constant in the same function will use the same address. IOW, this
```rust
const S: &str = "hello";
let x = S;
let y = S;
```
is conceptually equivalent to
```rust
let x = "hello";
let y = "hello";
```
and there is no de-duplication guarantee. That's in general how constants behave: we think of them as basically having their definition inlined at each usage site.

File: compiler/rustc_mir_transform/src/gvn.rs
Comment (oli-obk) [inline compiler/rustc_mir_transform/src/gvn.rs:257]: I don't understand this comment.

**Review by oli-obk — APPROVED** [review]

File: src/tools/miri/tests/pass/function_pointers.rs
Comment (RalfJung) [inline src/tools/miri/tests/pass/function_pointers.rs:85]: This is somewhat dubious. Certainly we want to test that we still successfully give them different addresses sometimes, e.g. by adding `black_box`.

But I am also not convinced that what GVN is doing here is sound. What is the argument for that?

Comment (RalfJung) [reply src/tools/miri/tests/pass/function_pointers.rs:85]: I guess if this is something specific to fn-item to fn-ptr coercions, then what the opt does is fine -- that coercion is allowed to deduplicate.

But please fix the test by adding `black_box` and still asserting `!=`. Let's not reduce test coverage unnecessarily.

**Pull request merged**
